planet/api/http.py

```diff
@@ -13,135 +13,176 @@
+                    await asyncio.sleep(wait_time)
         raise Exception('too many throttles, giving up')
 
-    def _do_request(self, request, **kwargs):
+    async def request(self, request, stream=False):
+        '''Submit a request with retry.'''
+        # TODO: retry will be provided in httpx v1 [1] with usage [2]
+        # 1. https://github.com/encode/httpcore/pull/221
+        # 2. https://github.com/encode/httpx/blob/
+        # 89fb0cbc69ea07b123dd7b36dc1ed9151c5d398f/docs/async.md#explicit-transport-instances # noqa
+        # TODO: if throttling is necessary, check out [1] once v1
+        # 1. https://github.com/encode/httpx/issues/984
+        return await self.retry(self._request, request, stream=stream)
+
+    async def _request(self, request, stream=False):
         '''Submit a request
 
-        :param :py:Class:`planet.api.models.Request` req: request to submit
-        :returns: :py:Class:`planet.api.models.Response`
+        :param request: Request to submit
+        :type request: planet.api.models.Request
+        :param stream: Get the body as a stream. Defaults to False.
+        :type stream: boolean, optional
+        :returns: response
+        :rtype: planet.api.models.Response
         '''
-        # TODO: I don't know where kwargs are used, maybe nowhere?
-        LOGGER.debug('about to submit request')
-        _log_request(request)
-
-        t = time.time()
-        http_resp = self._session.request(
-            request.method, request.url, data=request.data,
-            headers=request.headers, params=request.params)
-        LOGGER.debug('request took %.03f', time.time() - t)
-
-        resp = models.Response(request, http_resp)
-        resp.raise_for_status()
-        return resp
-
-
-class RedirectSession(Session):
-    '''This exists to override the existing behavior of requests that will
-    strip Authorization headers from any redirect requests that resolve to a
-    new domain. Instead, we'll keep headers if the redirect is a subdomain
-    and if not, extract the api-key from the header and add it to the url
-    as a parameter.
-    '''
-    def rebuild_auth(self, prepared_request, response):
-        existing_auth = prepared_request.headers.get('Authorization', None)
-        if existing_auth:
-            orig = response.request.url
-            redir = prepared_request.url
-            if not self._is_subdomain_of_tld(orig, redir):
-                prepared_request.headers.pop('Authorization')
-                key = re.match(r'api-key (\S+)', existing_auth)
-                if key:
-                    prepared_request.prepare_url(
-                        prepared_request.url, {
-                            'api_key': key.group(1)
-                        }
-                    )
+        http_resp = await self._client.send(request.http_request,
+                                            stream=stream)
+        return models.Response(request, http_resp)
+
+    def stream(self, request):
```

Comment:
Do we think that this happens often enough for this function? The reason I ask is because the pattern appears inconsistent and can be confusing. That is to say, encouraging users to `with planet.Session(...) as session: with planet.OrdersClient(session) as oc: ...` feels more consistent to me than `with ... as session: with session.stream(req): ...`. I'm curious about what benefit this provides over something like `with ... as session: with planet.Stream(session, req) as stream: ...`?

Reply:
This pattern follows the pattern used in [httpx](https://www.python-httpx.org/async/#streaming-responses), the library we are using for all API calls. So the primary benefit is not having to do a lot of implementation and add confusion to those familiar with httpx by changing the usage pattern. In reality, this is not a function the public should be using, it is used internally to manage receiving and writing information from the server (aka downloading an image). Your confusion here is valuable information. Is there a way we can use documentation to let users know more about this usage pattern?

Reply:
I'm curious, but it seems as though the call chain specifically looks like it goes from `Session.stream` to `APlanetStream` to `Session.request`. I feel like I'm missing something because calling `Session.request(..., stream=True)` appears, at first glance, to be able to replace those extra jumps/layers. I want to make sure I'm understanding enough of this to continue commenting on it. :)

Reply:
`APlanetStream` is a context manager, that's the value it adds. So in addition to calling `Session.request`, it is calling `response.aclose()` when the stream is no longer needed.

Reply:
Okay, so in that case it seems like:
```python
response = await session.request(request, stream=True)
body = StreamingBody(response)
return await body.write(...)
```
is the same as...
```python
async with session.stream(request) as response:
    body = StreamingBody(response)
    return await body.write(...)
```
right?

Reply:
agree with @jreiberkyle ... this is an async context manager, and its implementation and usage suggested here is in alignment with python+async norms. This is fine.